tsdb/index/index.go

```diff
@@ -1511,6 +1511,51 @@ func (r *Reader) LabelValues(name string, matchers ...*labels.Matcher) ([]string
 	return values, nil
 }
 
+// LabelNamesFor returns all the label names for the series referred to by IDs.
+// The names returned are sorted.
+func (r *Reader) LabelNamesFor(ids ...uint64) ([]string, error) {
+	// Gather offsetsMap the name offsetsMap in the symbol table first
+	offsetsMap := make(map[uint32]struct{})
+	for _, id := range ids {
+		offset := id
+		// In version 2 series IDs are no longer exact references but series are 16-byte padded
+		// and the ID is the multiple of 16 of the actual position.
+		if r.version == FormatV2 {
+			offset = id * 16
+		}
+
+		d := encoding.NewDecbufUvarintAt(r.b, int(offset), castagnoliTable)
+		buf := d.Get()
+		if d.Err() != nil {
+			return nil, errors.Wrap(d.Err(), "label names for")
+		}
+
+		offsets, err := r.dec.LabelNamesOffsetsFor(buf)
+		if err != nil {
+			return nil, storage.ErrNotFound
+		}
+		for _, off := range offsets {
+			offsetsMap[off] = struct{}{}
+		}
+	}
+
+	// lookup the unique symbols
+	names := make([]string, 0, len(offsetsMap))
+	for off := range offsetsMap {
+		name, err := r.lookupSymbol(off)
+		if err != nil {
+			return nil, errors.Wrap(err, "lookup symbol in label names for")
+		}
+		names = append(names, name)
+	}
+
+	// maybe it's more efficient to built the list of offsets first, then sort the uint32 offsets and then map to names?
+	// that could be an optimization if we measure it and makes any difference
```

Comment:
Number of names is usually not that big, so this way should be fine. It's an additional slice to sort the offsets, so might as well skip it for now. We can remove this comment.

Reply:
Removed here: https://github.com/prometheus/prometheus/pull/9083/commits/1982c8ea0eb1609289c7473bc843eb6fcaa1196c